Sorting Queries

Skn0tt opened the pull request:
Closes #8 

Love this plugin! I've started getting into Obsidian yesterday, and this plugin seems to fit my workflow pretty well.

This PR adds support for `sort by due` style directives. I intentionally flesh out the concrete syntax and how priorities / sorting by multiple fields should work. I'm pretty sure you have some great ideas on it @schemar.

TODO:
- [x] discuss concrete syntax
  - `sort by` / `order by` / something entirely different?
  - how to specify sorting by multiple fields? comma-separated / multiple sort-by-directives?
  - which fields should be sortable?
  - how to do ascending / descending?
- [x] write acceptance tests for agreed-upon syntax
- [x] implement!

Skn0tt commented:
this PR is intentionally broken up in small commits that tell a story, so please review them one-by-one :)

schemar commented:
OMG thank you @Skn0tt ! My time is very limited right now, but I will get to it as soon as I can. Please be patient :smiley:

schemar commented on src/Query.ts:
```diff
@@ -2,10 +2,13 @@ import chrono from 'chrono-node';
 
 import { Status, Task } from './Task';
 
+type Sorting = 'status' | 'due' | 'done';
```
Should also support `path`.

Skn0tt replied:
d869b95

schemar commented on src/Query.ts:
```diff
@@ -217,6 +227,16 @@ export class Query {
         }
     }
 
+    private readonly sortByRegexp = /sort by (status|due|done)/;
```
I prefer all regexes at the top to see at first glance what's possible.

Please add the caret at the beginning to prevent mix ups :)
```suggestion
    private readonly sortByRegexp = /^sort by (status|due|done)/;
```

Skn0tt replied:
#1 fixed in ca3d6dc

Re caret: the other regexes don't have a caret, and I don't think it's a problem to not have it b/c it does line-based matching anyway.
https://github.com/schemar/obsidian-tasks/blob/ca3d6dc496c4c2f58b3cc270ac082c3a175f60cb/src/Query.ts#L68-L68

I'd prefer to keep the caret usage consistent amongst the regexes, otherwise I'd be confused as to why one of them uses it and the others don't

Skn0tt replied:
Forget what I said about carets, I just merged in `main`. There they are ^^

Skn0tt replied:
49050bc

schemar commented on src/Sort.ts:
```diff
@@ -1,8 +1,46 @@
+            this.compareByPath,
+        ];
+
+        for (const sortProp of query.sorting.reverse()) {
+            switch (sortProp) {
+                case 'status':
+                    priorities.unshift(this.compareByStatus);
+                    break;
+                case 'due':
+                    priorities.unshift(this.compareByDueDate);
+                    break;
+                case 'done':
+                    priorities.unshift(this.compareByDoneDate);
+                    break;
+            }
+        }
```
This means you do a little more sorting than necessary, but I think that should be fine. You would sort by `status` twice, for example. But as I said, I think it's ok this way.

Skn0tt replied:
yeah, also thought about that. but since sorting works through the properties [one-by-one](https://github.com/Skn0tt/obsidian-tasks/blob/d869b95046dd6ce32dd2cc9bd1ec490ebe01c866/src/Sort.ts#L39-L39), double-sorting will happen only for *very* similar items. and since this is only about speed, not about sorting correctness (sorting twice by the same property won't change anything, no matter the sort order), I think it should be fine :D

schemar reviewed (commented):
To be honest, there is not much to complain or discuss :sweat_smile: 
I really like the approach and that you added tests. I mean, there could be more tests, but the fact that you added any at all I already like a lot :heart: 

The only thing missing would be an option to define "ascending" or "descending". What do you think?

Skn0tt commented:
Happy you like the PR! I addressed your comments. I'll add some tests for the following API decisions:

- `sort by` is the keyword
- to sort by multiple fields, write multiple lines of `sort by`. the higher-up, the more prioritised it is.

I'll open a second PR for ascending / descending, wanna keep this one lean.

Skn0tt commented:
Added tests! This PR is merge-ready.

schemar commented:
Please resolve conflicts and make sure to run `yarn lint` before pushing the changes. Thank you! :raised_hands:

WPum commented:
Can "sort by description" be added?

WPum commented:
How about adding a "sort by description"?

Skn0tt commented:
Good idea @WPum! added it in 3ba9f27

WPum commented:
Great!

schemar reviewed (approved):
Love it, thank you! :tada:
